PowerPointLabs/PowerPointLabs/DataSources/ColorDataSource.cs

```diff
@@ -1,515 +1,183 @@
+                    selectedColorValue = value;
+                    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("selectedColor"));
                 }
             }
         }
 
-        private HSLColor themeColorTwoValue;
-
-        public HSLColor ThemeColorTwo
-        {
-            get
-            {
-                return themeColorTwoValue;
-            }
-            set
-            {
-                if (value != this.themeColorTwoValue)
-                {
-                    this.themeColorTwoValue = value;
-                    OnPropertyChanged("themeColorTwo");
-                }
-            }
-        }
+        #endregion
 
-        private HSLColor themeColorThreeValue;
+        #region Constructors
 
-        public HSLColor ThemeColorThree
+        public ColorDataSource()
         {
-            get
-            {
-                return themeColorThreeValue;
-            }
-            set
-            {
-                if (value != this.themeColorThreeValue)
-                {
-                    this.themeColorThreeValue = value;
-                    OnPropertyChanged("themeColorThree");
-                }
-            }
+            recentColors = new ObservableCollection<HSLColor>();
+            favoriteColors = new ObservableCollection<HSLColor>();
+            recentColors.CollectionChanged += RecentColors_CollectionChanged;
+            favoriteColors.CollectionChanged += FavoriteColors_CollectionChanged;
+            ClearRecentColors();
+            ClearFavoriteColors();
         }
 
-        private HSLColor themeColorFourValue;
-
-        public HSLColor ThemeColorFour
-        {
-            get
-            {
-                return themeColorFourValue;
-            }
-            set
-            {
-                if (value != this.themeColorFourValue)
-                {
-                    this.themeColorFourValue = value;
-                    OnPropertyChanged("themeColorFour");
-                }
-            }
-        }
+        #endregion
 
-        private HSLColor themeColorFiveValue;
+        #region API
 
-        public HSLColor ThemeColorFive
+        public void AddColorToRecentColors(HSLColor color)
         {
-            get
-            {
-                return themeColorFiveValue;
-            }
-            set
+            int index = recentColors.IndexOf(color);
+            if (index == -1)
             {
-                if (value != this.themeColorFiveValue)
-                {
-                    this.themeColorFiveValue = value;
-                    OnPropertyChanged("themeColorFive");
-                }
+                index = recentColors.Count - 1;
             }
-        }
-
-        private HSLColor themeColorSixValue;
 
-        public HSLColor ThemeColorSix
-        {
-            get
-            {
-                return themeColorSixValue;
-            }
-            set
+            for (int i = index - 1; i > 0; i--)
```

Comment:
This line of code is causing the recent colors not to update properly. Should start from `index` instead of `index - 1`